_posts/2019-06-07-python-datascience-with-openfaas.md

```diff
@@ -0,0 +1,267 @@
+├── handler.py
+├── requirements.txt
+└── train.py
+```
+
+### Training
+Another point to note is the training data folder is also included here, `data/names` and a serialized model is also include `data/char-rnn-classification.pt`.
+
+This template is designed to run the training as part of the build process.  You can easily trigger the training using
+
+```bash
+python train.py
+```
+
+This will generate a new serialized model `data/char-rnn-classification.pt`.
+
+There are two methods for [saving PyTorch models][pytorch-saving-models]:
+
+1. save the entire model using `pickle`, or
+2. saving just the model `state_dict`.
+
+
+We have opted for the second method because it is more portable, the `pickle` method is sensitive to the Python environment and folder structure.
+
+```python
+torch.save(rnn.state_dict(), "data/char-rnn-classification.pt")
+```
+
+### Loading the model into function memory
+Using the `HTTP` mode in the [of-watchdog][of-watchdog-homepage] enables us to load this model into memory once and reuse it for multiple requests. We only need to load the model at the start of the `handler.py` implementation
+
+```python
+import json
+import os
+from pathlib import PurePath
+from typing import Any, List
+
+import torch
+from torch.autograd import Variable
+
+from .core import const, model, utils
+
+FUNCTION_ROOT = os.environ.get("function_root", "/root/function/")
+
+# init model
+RNN = model.RNN(const.N_LETTERS, const.N_HIDDEN, const.N_CATEGORIES)
+# fill in weights
+RNN.load_state_dict(
+    torch.load(str(PurePath(FUNCTION_ROOT, "data/char-rnn-classification.pt")))
+)
+
+
+def predict(line: str, n_predictions: int = 3) -> List[Any]:
+    """omitted for brevity"""
+
+def handle(req: bytes) -> str:
+    """handle a request to the function
+    Args:
+        req (bytes): request body
+    """
+
+    if not req:
+        return json.dumps({"error": "No input provided", "code": 400})
+
+    name = str(req)
+    output = predict(name)
+
+    return json.dumps(output)
+```
+
+When a function instance is instantiated, e.g. during scaling or deployment, this file is parsed and the model is loaded into memory. This will happen exactly once because the HTTP mode loads this handler as a very small background web sever.  The original forking mode in the watchdog would instead load this file for every invocation. When loading models, this creates additional latency.
+
+This template uses Flask to power the background web server. The `handle` can return any value that [Flask would accept][flask-return-values].  This provides a lot of flexibility to control the response. For example, you could instead set the error status code and body like this
+
+```python
+return json.dumps({"error": "No input provided"}), 400
+```
+
+### Deployment
+This template is designed to bundle the pre-trained model into the final Docker image. This means that the deployment steps look like this
+
+```bash
+cd classify && python train.py && cd -
```

Comment:
`&& cd -`?

Reply:
`cd -` sends you back to your previous directory, you need to go back to the project root to run the build

Reply:
i will use `cd ..` instead